californium-core/src/main/java/org/eclipse/californium/core/network/Matcher.java

```diff
@@ -353,18 +367,24 @@ public Exchange receiveResponse(Response response) {
 				// See issue #275
 				return null;
 			}
-
-			// we have received a Response matching the token of an ongoing Exchange's Request
-			// according to the CoAP spec (https://tools.ietf.org/html/rfc7252#section-4.5),
-			// message deduplication is relevant for CON and NON messages only
-			if ((response.getType() == Type.CON || response.getType() == Type.NON) &&
-					deduplicator.findPrevious(idByMID, exchange) != null) {
-				LOGGER.info("Duplicate response for open exchange: "+response);
-				response.setDuplicate(true);
-			} else {
-				idByMID = new KeyMID(exchange.getCurrentRequest().getMID(), null, 0);
-				exchangesByMID.remove(idByMID, exchange);
-				if (LOGGER.isLoggable(Level.FINE)) LOGGER.fine("Closed open request with "+idByMID);
+			
+			if (exchange.getCurrentRequest().isMulticast()) {
+			    // keep exchange open because we might receive additional responses
+			    LOGGER.log(Level.FINER, "Received response for multicast request: {0}", response);
+			}
+			else {
+				// we have received a Response matching the token of an ongoing Exchange's Request
+				// according to the CoAP spec (https://tools.ietf.org/html/rfc7252#section-4.5),
+				// message deduplication is relevant for CON and NON messages only
+				if ((response.getType() == Type.CON || response.getType() == Type.NON) &&
+						deduplicator.findPrevious(idByMID, exchange) != null) {
+					LOGGER.info("Duplicate response for open exchange: "+response);
+					response.setDuplicate(true);
+				} else {
+					idByMID = new KeyMID(exchange.getCurrentRequest().getMID(), null, 0);
+					exchangesByMID.remove(idByMID, exchange);
```

Comment:
As far as I know, exchangesByMID are only relevant for ACK or RST. 
ACK are not used in multicast and using RST would cause RFC7252, 8.1, page 65:

> the sender MUST avoid using a
> Message ID that is also still active from this endpoint with any
> unicast endpoint that might receive the multicast message.

which would be hard to achieve when consider a server with more message load.
The 1.0 branch is using central "MID counter" for all clients, which limits the overall traffic, but grants the above.
The 2.0 branch therefore use a "per connection MID counter", so there we would need the ensure that explicitly.
(FMPOV multicast seems to be a more rare use-case then to sending more messages).

So do you use RST? If not, why do you need to change the response processing? Which issues did you observe, if the responses are processed usual?

Reply:
No I didn't use the RST and you are right again : with the filter in receiveResponse of CoapStack.java I didn't need to filter the multicast response again in receiveResponse of Matcher.java.